app/Http/Controllers/DemandController.php

```diff
@@ -0,0 +1,334 @@
+            $response->headers->set("Access-Control-Expose-Headers", "X-Adshares-Cid, X-Adshares-Lid");
+        }
+
+        if ($request->getRealMethod() == 'OPTIONS') {
+            $response->headers->set('Access-Control-Max-Age', 1728000);
+            return $response;
+        }
+
+        $isIECompat = $request->query->has("xdr");
+
+        if ($banner->creative_type == 'html') {
+            $mime = "text/html";
+        } else {
+            $mime = "image/png";
+        }
+
+        $tid = Utils::attachTrackingCookie(config('app.adserver_secret'), $request, $response, $banner->creative_sha1, $banner->updated_at);
+
+        $response->setCallback(function () use ($response, $banner, $isIECompat) {
+            if (!$isIECompat) {
+                echo $banner->creative_contents;
+                return;
+            }
+
+            $headers = [];
+            foreach ($response->headers->allPreserveCase() as $name => $value) {
+                if (strpos($name, "X-") === 0) {
+                    $headers[] = "$name:" . implode(",", $value);
+                }
+            }
+            echo implode("\n", $headers) . "\n\n";
+            echo base64_encode($banner->creative_contents);
+        });
+
+        $logIp = bin2hex(inet_pton($request->getClientIp()));
+        $cid = Utils::createTrackingId(config('app.adserver_secret'));
+
+        $log = new EventLog();
+        $log->banner_id =$banner->id;
+        $log->cid = Utils::getRawTrackingId($cid);
+        $log->tid = Utils::getRawTrackingId($tid);
+        $log->ip = $logIp;
+        $log->event_type = "request";
+        $log->save();
+
+        $response->headers->set('X-Adshares-Cid', $cid);
+        $response->headers->set('X-Adshares-Lid', $log->id);
+
+        if (! $response->isNotModified($request)) {
+            $response->headers->set('Content-Type', ($isIECompat ? 'text/base64,' : '') . $mime);
+        }
+        return $response;
+    }
+
+    public function viewScript(Request $request)
+    {
+        $params = [json_encode($request->getSchemeAndHttpHost())];
+
+        $jsPath = config('app.env') == 'production' ? resource_path('assets/js/tmp-copy/view.x.js') : resource_path('assets/js/tmp-copy/view.min.js');
+
+        $response = new StreamedResponse();
+        $response->setCallback(function () use ($jsPath, $request, $params) {
+            echo str_replace([
+              "'{{ ORIGIN }}'",
+            ], $params, file_get_contents($jsPath));
+        });
+
+        $response->headers->set('Content-Type', 'text/javascript');
+
+        $response->setCache(array(
+          'etag' => md5(md5_file($jsPath) . implode(':', $params)),
+          'last_modified' => new \DateTime('@' . filemtime($jsPath)),
+          'max_age' => 3600 * 24 * 30,
+          's_maxage' => 3600 * 24 * 30,
+          'private' => false,
+          'public' => true
+        ));
+
+        if (! $response->isNotModified($request)) {
+            // TODO: ask Jacek
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)